source/air/src/singular_manager.rs

```diff
@@ -2,28 +2,31 @@ use std::io::{BufRead, BufReader, Write};
 use std::process::{ChildStdin, ChildStdout};
 use std::sync::mpsc::{channel, Receiver, Sender};
 
+// this is only used in print command
+pub const DONE: &str = "<<Singular DONE>>";
+
 pub struct SingularManager {
     pub singular_executable_name: String,
 }
 
 pub struct SingularProcess {
-    requests: Sender<Vec<u8>>,
+    requests: Option<Sender<Vec<u8>>>,
     singular_pipe_stdout: BufReader<ChildStdout>,
+    child: std::process::Child,
 }
 
-/// A separate thread writes data to the Singular solver over a pipe.
-/// (Rust's documentation says you need a separate thread; otherwise, it lets the pipes deadlock.)
-pub(crate) fn writer_thread(requests: Receiver<Vec<u8>>, mut smt_pipe_stdin: ChildStdin) {
+/// this is similar to writer_thread in smt_process.rs
+/// the way we ask for acknowledgement is slightly different
+fn singular_writer_thread(requests: Receiver<Vec<u8>>, mut singular_pipe_stdin: ChildStdin) {
     while let Ok(req) = requests.recv() {
-        smt_pipe_stdin
+        singular_pipe_stdin
             .write_all(&req)
-            .and_then(|_| writeln!(&smt_pipe_stdin))
-            .and_then(|_| smt_pipe_stdin.flush())
-            // The Singular process could die unexpectedly.  In that case, we die too:
-            // TODO: https://github.com/verus-lang/verus/pull/730
+            // ask for acknowledgement
+            .and_then(|_| writeln!(&singular_pipe_stdin, "print (\"{}\");", DONE))
+            .and_then(|_| singular_pipe_stdin.flush())
             .expect("IO error: failure when sending data to Singular process across pipe");
     }
-    // Exit when the other side closes the channel
+    // singular_pipe_stdin.write_all(b"quit;\n").expect("IO error: failure quitting Singular process");
```

Comment:
Why is this commented out?

Reply:
Ah good catch, I think I was debugging some odd IO error, and forgot to turn this back on. Thanks for noticing.